### PR TITLE
Dev

### PR DESCRIPTION
**Pull request by ilovec0ncernn**
Pull request (code review)

### CODE REVIEW

File: Test.cpp [hunk 1]
```diff
@@ -0,0 +1,60 @@
...
+    Network net(rng);
+    net.AddLayer(784, 128, &relu, rng);  // hidden-1 ReLU
+    net.AddLayer(128, 10, &sm, rng);     // hidden-2 Softmax
+    net.AddLayer(10, 10, &id, rng);      // выход логиты
+
+    // обучение
+    TrainConfig tcfg;
+    tcfg.epochs = cfg.epochs;
+    tcfg.batch_size = cfg.batch_size;
+    tcfg.lr = cfg.lr;
```
Comment (DimaTrushin) [inline Test.cpp:39]: 1) А чем у тебя TrainConfig отличается от TestConfig?
2) Это должно быть одна строчка, что-то такое:
```cpp
TrainConfig tcfg = TrainConfig::FromTestConfig(cfg);
```

File: Test.cpp [hunk 2]
```diff
@@ -0,0 +1,60 @@
...
+    MSE mse;
```
Comment (DimaTrushin) [inline Test.cpp:40]: А это где используется?

File: Test.cpp [hunk 3]
```diff
@@ -0,0 +1,60 @@
...
+    CrossEntropyWithLogits celoss;
+    net.Train(X_train, y_train, X_test, y_test, tcfg, celoss);
+
+    // итоговые метрики
+    Accuracy acc_metric;
+    CrossEntropyMetric ce_metric;
```
Comment (DimaTrushin) [inline Test.cpp:46]: а почему метрики зависят от вида лосса?

File: Test.cpp [hunk 4]
```diff
@@ -0,0 +1,60 @@
...
+    net.AddLayer(10, 10, &id, rng);      // выход логиты
+
+    // обучение
+    TrainConfig tcfg;
+    tcfg.epochs = cfg.epochs;
+    tcfg.batch_size = cfg.batch_size;
+    tcfg.lr = cfg.lr;
+    MSE mse;
+    CrossEntropyWithLogits celoss;
+    net.Train(X_train, y_train, X_test, y_test, tcfg, celoss);
```
Comment (DimaTrushin) [inline Test.cpp:42]: Было бы интересно собирать отчет по обучению. Например смотреть на разные метрики, как они меняются во время обучения с ходом эпох.

File: Test.cpp [hunk 5]
```diff
@@ -0,0 +1,60 @@
...
+#include <iostream>
+
+namespace nn {
+
+static Matrix TakeRows(const Matrix& M, int n) {
+    if (n < 0 || n >= M.rows()) {
+        return M;
+    }
+    return M.topRows(n);
+}
```
Comment (DimaTrushin) [inline Test.cpp:13]: Я понимаю, что это костыль вокруг библиотечного кода, но выглядит так, что этому тут не место. Но пока не знаю, куда это.

File: Network.h [hunk 6]
```diff
@@ -0,0 +1,39 @@
...
+    void Train(const Matrix& X, const Matrix& Y, const Matrix& X_val, const Matrix& Y_val, const TrainConfig& cfg,
+               LossFunction& loss);
+
+    // предсказания
+    Vector PredictOne(const Vector& x) const;  // прямой прогон одного объекта
+    Matrix Predict(const Matrix& X) const;     // прогон матрицы объектов (построчно)
+
+   private:
+    std::vector<Layer> layers_;
+    std::mt19937_64 shuffle_eng_;
```
Comment (DimaTrushin) [inline Network.h:36]: 1) Зачем тут запоминать генератор рандома? Во-первых, это слишком низкоуровневая вещь. Ты не должен генератором пользоваться напрямую. У тебя должен быть класс class Random, который инкапсюлирует всю работу с рандомом и он уже хранит генератор и он уже тебе предоставляет возможность сгенерировать случайные вещи.
2) Генератор видимо нужен в конструкторе и для обучения. Но тогда его и надо сообщать в конструкоре и во время обучения.
Тут этого быть не должно.

File: Network.h [hunk 7]
```diff
@@ -0,0 +1,39 @@
...
+class Network {
+   public:
+    explicit Network(RNG& rng);
+
+    // конструируем архитектуру нейросети
+    void AddLayer(Index in_dim, Index out_dim, const ActivationFunction* sigma, RNG& rng);
+
+    // обучение и вывод метрик в std::cout
+    void Train(const Matrix& X, const Matrix& Y, const Matrix& X_val, const Matrix& Y_val, const TrainConfig& cfg,
+               LossFunction& loss);
```
Comment (DimaTrushin) [inline Network.h:28]: Почепму у тебя loss принимается не по константной ссылке? ты что имеешь state у loss?

File: Network.h [hunk 8]
```diff
@@ -0,0 +1,39 @@
...
+    int batch_size = 64;
+    float lr = 0.05f;
+};
+
+class Network {
+   public:
+    explicit Network(RNG& rng);
+
+    // конструируем архитектуру нейросети
+    void AddLayer(Index in_dim, Index out_dim, const ActivationFunction* sigma, RNG& rng);
```
Comment (DimaTrushin) [inline Network.h:24]: Плохой интерфейс.
1) Во-первых, с ним легко ошибиться. Вот что значит
```cpp
net.AddLayer(100, 200, ..., ...);
```
Что из этого размер входа, а что размер выхода? Но еще хуже, такой интерфейс подразумевает, что ты должен уже знать, какого размера слой уже лежит внутри, то есть ты можешь сделать так
```cpp
net.AddLayer(100, 200, ..., ...);
net.AddLayer(100, 200, ..., ...);
```
И все, у тебя размеры слоев не согласованы. Ты по сути попытался сделать Builder паттерн.
Я понимаю, что ты не можешь принимать только один размер, потому что в случае добавления самого первого слоя у нас нет начального размера. Тут есть два варианта:
1) сделать отдельно метод для добавления начального слоя (и если мы его вызываем два раза, то мы перезаписываем начальный слой). А методы для добавления остальных слоев лишь принимают размер выхода.
2) сделать отдельный класс Builder, который хранит пустую нейросеть и размер предыдущего слоя. В билдере ты можешь хранить любую промежуточную информацию.

Comment (DimaTrushin) [inline Network.h:24]: Какого хрена у тебя ActivationFunction принимается по константному указателю? Что за дичь? Почему не константная ссылка?

File: Network.cpp [hunk 9]
```diff
@@ -0,0 +1,108 @@
...
+
+        for (int i = 0; i < n; i += b) {
+            const int r = std::min(b, n - i);
+
+            for (auto& L : layers_) {
+                L.ZeroGrad();
+            }
+
+            for (int j = 0; j < r; ++j) {
+                Index idx = order[i + j];
```
Comment (DimaTrushin) [inline Network.cpp:65]: 4 вложенных scope. Даже в ядре Linux запрещено использовать больше 3.

File: Network.cpp [hunk 10]
```diff
@@ -0,0 +1,108 @@
...
+                y_true.maxCoeff(&yi);
+                logits.maxCoeff(&pi);
+                correct += (yi == pi);
+                ++seen;
+
+                Vector grad = loss.Gradient(y_true, logits);  // dL/d(logits)
+
+                // backpropagation
+                grad = layers_.back().BackwardDy(grad);
+                for (int l = static_cast<int>(layers_.size()) - 2; l >= 0; --l) {
```
Comment (DimaTrushin) [inline Network.cpp:87]: Никогда НЕ используй static_cast в бизнес логике. У тебя интерфейсы классов предоставляют данные не в том формате, в котором они их ожидают, и потому ты занимаешься такой хуйней как конвертация данных напрямую.
1) Во-первых, начиная с c++20 есть знаковый размер std::ssize.
2) Во-вторых, можно написать свою ssize и звать ее тут вместо вот этой хуйни.

File: Network.cpp [hunk 11]
```diff
@@ -0,0 +1,108 @@
...
+                    x = L.Forward(x);
+                }
+                Vector logits = x;
+
+                // train accuracy
+                int yi = 0, pi = 0;
+                y_true.maxCoeff(&yi);
+                logits.maxCoeff(&pi);
+                correct += (yi == pi);
+                ++seen;
```
Comment (DimaTrushin) [inline Network.cpp:81]: Это должна быть одна функция с нормальным названием, которая говорит, что за хуйня у тебя тут происходит.

File: ActivationFunctions.h [hunk 12]
```diff
@@ -0,0 +1,33 @@
...
+   public:
+    Vector Forward(const Vector& z) const override;
+    Vector Derivative(const Vector& z) const override;
+};
+
+class Softmax : public ActivationFunction {
+   public:
+    Vector Forward(const Vector& z) const override;
+    Vector Derivative(const Vector& z) const override;
+    Vector BackwardFromDy(const Vector& y, const Vector& dL_dy) const;
```
Comment (DimaTrushin) [inline ActivationFunctions.h:30]: На самом деле во всех Activation Function тебе нужны только методы forward и backward. И можно тогда обойтись без виртуальных методов:
```cpp
class Activation {
  using ForwardSig = Vector(const Vector&);
  using BackwardSig = Vector(const Vector&, const Vector&);
public:
  Vector forward(const Vector& x) const {
    return forward_(x);
  }
  Vector backward(const Vector& y, const Vector& u) const {
    return backward_(y, u);
  }
static Activation ReLU() {
    return Activation(/*Relu forward*/, /Relu backward/);
  }
...
private:
  std::function<ForwardSig> forward_;
  std::function<BackwardSig> backward_;
};
```
И все. И не надо ебать мозги вот этой всей хуйней со указателями на хуй пойми что и dynamic cast-ами.

File: ActivationFunctions.cpp [hunk 13]
```diff
@@ -0,0 +1,45 @@
...
+
+Vector Softmax::Forward(const Vector& z) const {
+    const Scalar m = z.maxCoeff();
+    const Vector e = (z.array() - m).exp().matrix();
+    const Scalar s = e.sum();
+    return e / s;
+}
+
+Vector Softmax::Derivative(const Vector& z) const {
+    Vector d(z.size());
```
Comment (DimaTrushin) [inline ActivationFunctions.cpp:33]: Вектор -- не контейнер. Не надо создавать его пустым и потом заполнять. Вместо этого ты можешь сделать поэлементные операции, как ты делал в других функциях. И получится без цикла.

File: ActivationFunctions.cpp [hunk 14]
```diff
@@ -0,0 +1,45 @@
...
+
+Vector ReLU::Forward(const Vector& z) const {
+    return z.cwiseMax(0.0f);
+}
+
+Vector ReLU::Derivative(const Vector& z) const {
+    Vector d(z.size());
+    for (Index i = 0; i < z.size(); ++i) {
+        d[i] = (z[i] > 0.0f) ? 1.0f : 0.0f;
+    }
```
Comment (DimaTrushin) [inline ActivationFunctions.cpp:13]: Убери цикл с помощью поэлементных операций.

File: InputDataset.h [hunk 15]
```diff
@@ -0,0 +1,23 @@
+#pragma once
+
+#include <string>
+
+#include "Alias.h"
+
+#ifndef NN_MNIST_DIR
+#define NN_MNIST_DIR "external_submodules/mnist"
+#endif
```
Comment (DimaTrushin) [inline InputDataset.h:9]: Вот такое должно быть зашито в систему сборки. Это тоже магическая константа. Убрать отсюда. Лучше тут сообщить об ошибке, что такой макрос не объявлен.

File: InputDataset.h [hunk 16]
```diff
@@ -0,0 +1,23 @@
...
+
+namespace nn {
+
+struct Split {
+    Matrix X_train, y_train;
```
Comment (DimaTrushin) [inline InputDataset.h:14]: Не объявляй переменные в одну строку.

File: Test.cpp (hunk 4 above)
Comment (ilovec0ncernn) [reply Test.cpp:46]: TODO

File: Network.cpp [hunk 17]
```diff
@@ -0,0 +1,108 @@
...
+#include <iostream>
+#include <numeric>
+
+namespace nn {
+
+Network::Network(RNG& rng) {
+    // число из нашего EigenRand генератора используем как seed для std::shuffle
+    auto seed = static_cast<std::mt19937_64::result_type>(rng.gen());
+    shuffle_eng_.seed(seed);
+}
```
[the remark this answers is not in the document]
Comment (ilovec0ncernn) [reply Network.cpp]: тут я вроде что-то перехуячил, но я не думаю, что сделал лучше))
